File: src/helpers/travelling-salesman.js
/**
 * Checks to see if you are at a dead end of if you have further nodes you can visit.
 * 
 * @param {array} row 
 * @returns {boolean} 
 */
function noOptionsAvailable(row){
    let deadEnd = true;

    row.forEach(cell => {
        if (cell !== 0){
            deadEnd = false;
        }
    })

    return deadEnd;
}

/**
 * For given array of nodes and function creating matrix of variables, it displays the shortest path
 * 
 * @param {array} nodes 
 * @param {function} startMatrix 
 * @returns {array}
 */
export function shortestPath(nodes, startMatrix){
    let shortestPath = [],
        shortestDistance,
        matrix   = startMatrix(),
        order    = [],
        distance = 0;

    nodes.forEach((node, index)=>{
        /* initial distance is 0 as we haven't gone anywhere yet */
        addOption(0, index);
        addNextNode(0, index);

        /* add the next node to the list */
        function addNextNode(option, index){
            /* if node has nowhere to go check and go back */
            if(noOptionsAvailable(matrix[index])){
                /* add and reset */
                updateShortest(order, distance);
                goBackNode(option, index);
            } else {
                /* for each node linked add and check the next node */
                matrix[index].forEach((option, index)=>{
                    if (option !== 0){
                        addOption(option, index);
                        addNextNode(option, index);
                    } 
                });
                /* when cycle finished for node, go back to the previous node */
                goBackNode(option, index);
            }
        } 

        /* add to current path and distance */
        function addOption(option, index) {
            matrix.forEach(row => row[index] = 0);
            order.push(nodes[index]);
            distance += option;
        }

        /* return to previous node to try different paths */
        function goBackNode(option, index) {
            /* add back last row to matrix */
            var orig = startMatrix();
            for(let i=0; i<orig.length ; i++){
                matrix[i][index] = orig[i][index];
            }

            /* remove node from order and distance */
            order.splice([order.length-1]);
            distance -= option;
        }

        /* update shortest path */
        function updateShortest(order, distance) {
            if (order.length === matrix.length && (shortestDistance === undefined || distance < shortestDistance)){
                shortestPath     = [...order];
                shortestDistance = distance;
            }
        }

    });

    return {path : shortestPath.join(', '), distance : shortestDistance};
}

File: src/ruby/travelling-salesman.rb
class TravellingSalesman
    def initialize(nodes, startMatrix)
        @nodes = nodes
        @startMatrix = startMatrix
        @shortestPath = []
        @shortestDistance
        @order = []
        @matrix = createStartMatrix
        @distance = 0
    end

    ##
    # shortest path function
    # need to fix matrix override
    def shortestPath
        @nodes.each_with_index do |_, nodeIndex|
          addOption(0, nodeIndex)
          addNextNode(0, nodeIndex)
        end

        return {
            distance: @shortestDistance,
            path: @shortestPath.join(", ")
        }
    end

    # add option to current order and distance
    private
    def addOption(distance, index)
        @matrix.each {|row| row[index] = 0}

        @order << @nodes[index]
        @distance += distance
    end

    # goes to next node to add to route
    def addNextNode(distance, index)
        if (noOptionsAvailable(@matrix[index]))
            updateShortest
            goBackNode(distance, index)
        else
            @matrix[index].each_with_index do |option, index|
              if (option != 0)
                  addOption(option, index)
                  addNextNode(option, index)
              end
            end
            goBackNode(distance, index)
        end
    end

    # check to see if at a dead end
    def noOptionsAvailable(row)
        row.none? {|cell| cell != 0}
    end

    # go back to previous node to go down other paths from that node
    def goBackNode(distance, index)
        original  = createStartMatrix
        original.length.times {|i| @matrix[i][index] = original[i][index]}

        @order.pop
        @distance -= distance
    end

    ##
    # if the current order is valid and the distance is shorter
    # update the shortest path and distance
    def updateShortest
        if (@order.length == @startMatrix.length && (@shortestDistance.nil? || @distance < @shortestDistance ))
            @shortestPath = @order.dup
            @shortestDistance = @distance
        end
    end

    ##
    # create immutable start matrix
    def createStartMatrix
        @startMatrix.map{|row| row.dup}
    end

end

# change to matrix
startMatrix = [
        [0,  3,  2,  5,  0,  0,  0,  0],
        [3,  0,  3,  0,  3,  0,  0,  0],
        [2,  3,  0,  2,  5,  3,  7,  0],
        [5,  0,  2,  0,  0,  6,  0,  0],
        [0,  3,  5,  0,  0,  0,  4,  0],
        [0,  0,  3,  6,  0,  0,  0,  1],
        [0,  0,  7,  0,  4,  0,  0,  2],
        [0,  0,  0,  0,  0,  1,  2,  0]
    ]

nodes = ["1","2","3","4","5","6","7","8"]
travellingSalesman = TravellingSalesman.new(nodes, startMatrix)
puts travellingSalesman.shortestPath
